sys/proc.c

```diff
@@ -53,4 +60,148 @@ proc_t *proc_find(pid_t pid) {
   return p;
 }
 
+void proc_exit(int exitstatus) {
+  /* NOTE: This is a significantly simplified implementation! We currently
+     assume there is only one thread in a process. */
+
+  thread_t *td = thread_self();
+  proc_t *p = td->td_proc;
+  assert(p);
+
+  WITH_MTX_LOCK (&p->p_lock) {
+
+    assert(p->p_nthreads == 1);
+    /* If the process had other threads, we'd need to wake the sleeping ones,
+       request all of them except this one to call thread_exit from
+       exc_before_leave (using a TDF_? flag), join all of them to wait until
+       they terminate. */
+
+    /* XXX: If this process has any unwaited zombie children, assign them for
+       adoption by pid1 (init), who will wait for them. */
+
+    /* Clean up process resources. */
+    vm_map_delete(p->p_uspace);
+    fdtab_release(p->p_fdtable);
+
+    /* Record some process statistics that will stay maintained in zombie
+       state. */
+    p->p_exitstatus = exitstatus;
+
+    /* Turn the process into a zombie. */
```

Comment:
I don't understand. The exact order of actions in this function does not matter much, as it all happens with `p->p_lock` locked.